llama.cpp

```diff
@@ -8099,6 +8113,32 @@ struct llm_build_context {
                     Vcur = ggml_add(ctx0, Vcur, model.layers[il].bv);
                     cb(Vcur, "Vcur", il);
                 }
+                if (model.layers[il].attn_q_norm) {
+                   Qcur = ggml_view_3d(ctx0, Qcur, n_embd_head, n_head, n_tokens,
+                           ggml_element_size(Qcur) * n_embd_head,
+                           ggml_element_size(Qcur) * n_embd_head * n_head,
+                           0);
+                   cb(Qcur, "Qcur", il);
+                   Kcur = ggml_view_3d(ctx0, Kcur, n_embd_head, n_head_kv, n_tokens,
+                           ggml_element_size(Kcur) * n_embd_head,
+                           ggml_element_size(Kcur) * n_embd_head * n_head_kv,
+                           0);
+                   cb(Kcur, "Kcur", il);
+
+
+                   Qcur = llm_build_norm(ctx0, Qcur, hparams,
+                           model.layers[il].attn_q_norm,
+                           NULL,
+                           LLM_NORM, cb, il);
+                   cb(Qcur, "Qcur", il);
+
+                   Kcur = llm_build_norm(ctx0, Kcur, hparams,
+                           model.layers[il].attn_k_norm,
+                           NULL,
+                           LLM_NORM, cb, il);
+                   cb(Kcur, "Kcur", il);
+                }
```

Comment:
To simplify these a bit, the `ggml_reshape_3d` that's done before the rope calculation below could be moved before the q and k layernorms, and then the views here won't be necessary.

Also, maybe separate the k layernorm into its own condition instead of unconditionally applying it when the q layernorm weight is present.